File: PRG/TEST_KIT/완전탐색/피로도.js
// https://school.programmers.co.kr/learn/courses/30/lessons/87946

function solution(k, dungeons) {
  let answer = -1;

  goToD(k, [], 0);

  function goToD(cur, visited, cnt) {
    for (let i = 0; i < dungeons.length; i++) {
      if (!visited.includes(i) && cur >= dungeons[i][0]) {
        answer = Math.max(answer, cnt + 1);
        goToD(cur - dungeons[i][1], [...visited, i], cnt + 1);
      }
    }
  }

  return answer;
}

File: PRG/TEST_KIT/완전탐색/피로도.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/87946?language=cpp

#include <string>
#include <vector>

using namespace std;

vector<vector<int>> DG;
vector<bool> visited;
int answer = 0;

void dfs(int left, int cnt){
    answer = max(answer, cnt);
    
    for(int i=0; i < DG.size(); i++){
       if(DG[i][0] <= left && !visited[i]){
           visited[i] = true;
           dfs(left - DG[i][1], cnt+1);
           visited[i] = false;
       }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    DG = dungeons;
    visited.resize(dungeons.size(), false);
    
    dfs(k, 0);
    
    return answer;
}

